Fixes gh-2113 Update LuaJIT memory profiler description

pgulutzan opened the pull request:
There are many changes in addition to the requirements for the 2.8.1 change, mostly minor things like American spelling and different word order, and the number of iterations in the loop in the test.lua example.

There are two items that I'm especially unsure of.

doc/book/admin/server_introspection.rst -- this has a reference to a non-memory profiler, with a dead link to the luatex.org site. I don't propose to change it, but wonder whether -- maybe as a separate issue -- something should be added to this paragraph about Tarantool's profiling.

Profiler displays contain tabs. Eugene Osintsev replaced them with spaces. I didn't. Which way is better?

Buristan commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -514,15 +552,42 @@ the profiling report becomes much prettier:
+
+@<filename>:<line number> holds <number of still reachable bytes> bytes: <number of allocation events> allocs, <number of deallocation events> frees
+
+Sometimes, especially when there are many ``collectgarbage()`` calls,
```
It's not only caused by `collectgarbage()` call. When there are **really tons** of allocated and freed objects (for example, if a program runs a lot of time, or GC pressure is really high for a chunk to profile). When it happens, a profiler output becomes unreadable, if we want to understand what "lines of code" (i.e. Lua objects) still holds Lua memory.

Buristan commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -52,7 +52,7 @@ Below is a chunk of simple Lua code named ``test.lua`` to illustrate this.
 
     local t = {}
-    for i = 1, 1e5 do
+    for i = 1, 1e4 do
```
Side note: Nice [typo](https://github.com/tarantool/doc/pull/2251/files#diff-9af01d5422ec57beeeb70a8daa3a3231f4f457e4b7145439ebed8d49dc210138R154) catch!

Buristan reviewed (approved):
Hi, Peter!

Thanks for the patch!

The single nit:
The part about `collectgarbage()` should be slightly changed, because a large amount of deallocations can be caused when there are a lot of fast-dying objects are created by the program. May be the part about `collectgarbage()` calls too stands out. But the tons of dellocations can be obscured without any call of collectgarbage()`. And vise versa, many calls to `collectgarbage()` may no free any of Lua memory. Otherwise, LGTM!

Buristan commented:
> doc/book/admin/server_introspection.rst -- this has a reference to a non-memory profiler, with a dead link to the luatex.org site. I don't propose to change it, but wonder whether -- maybe as a separate issue -- something should be added to this paragraph about Tarantool's profiling.

Don't sure about it, but agree that it should be done as a separate issue.

> Profiler displays contain tabs. Eugene Osintsev replaced them with spaces. I didn't. Which way is better?

I suppose, that spaces are better in this case (for documentation). It helps to avoid confusing with different tab settings. I'm OK with both variants.

pgulutzan commented:
Okay, I removed the collectgarbage() emphasis by changing to
"
Sometimes a program can cause many deallocations, so
the DEALLOCATION section can become large, so the display is not easy to read.
"
and I changed the tabs to spaces.

NickVolynkin commented:
Hi, @pgulutzan. Just found this PR. My apologies, I somehow skipped it. Will now pass it to translators.

NickVolynkin commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -61,38 +61,40 @@ Below is a chunk of simple Lua code named ``test.lua`` to illustrate this.
     local stp, err = misc.memprof.stop()
 
-Starting profiler in Lua code:
+The Lua code for starting the profiler -- as in line 3 in the test.lua example above -- is:
```
```suggestion
The Lua code for starting the profiler --- as in line 3 in the test.lua example above -- is:
```
I believe it should be an m-dash, so three dashes.

NickVolynkin commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -129,73 +131,90 @@ Parsing binary profile and generating profiling report
-    @test.lua:0, line 10: 1  32      0
-    @test.lua:0, line 8: 1  20      0
-
+    @test.lua:14: 10000 events  +50240518 bytes -0 bytes
```
Now it looks much better, more understandable. 👍

NickVolynkin commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -129,73 +131,90 @@ Parsing binary profile and generating profiling report
-* DEALLOCATIONS.
+* REALLOCATIONS
+* DEALLOCATIONS
+* HEAP SUMMARY (described later in :ref:`The heap summary and the --leak-only option<luajit_memprof_leak_only>`)
```
There must be a space between the link text and target
```suggestion
* HEAP SUMMARY (described later in :ref:`The heap summary and the --leak-only option <luajit_memprof_leak_only>`)
```

NickVolynkin commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -129,73 +131,90 @@ Parsing binary profile and generating profiling report
-    is set to ``0``. See the :ref:`code chunk above<profiler_usage_example01>`
+    is declared. In some cases, allocations are attributed not to
+    the declared function but to the main chunk -- in that case, the <function_line>
+    is set to ``0``. See the :ref:`tarantool test.lua chunk above<profiler_usage_example01>`
```
same here, needs a space
```suggestion
    is set to ``0``. See the :ref:`tarantool test.lua chunk above <profiler_usage_example01>`
```

patiencedaur replied:
```suggestion
    is set to ``0``. See the :ref:`test.lua chunk above <profiler_usage_example01>`
```

NickVolynkin commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -129,73 +131,90 @@ Parsing binary profile and generating profiling report
-    the declared function but to the main chunk. In this case, the <function_line>
-    is set to ``0``. See the :ref:`code chunk above<profiler_usage_example01>`
+    is declared. In some cases, allocations are attributed not to
+    the declared function but to the main chunk -- in that case, the <function_line>
```
```suggestion
    the declared function but to the main chunk --- in that case, the <function_line>
```

NickVolynkin commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -61,38 +61,40 @@ Below is a chunk of simple Lua code named ``test.lua`` to illustrate this.
+If the operation succeeds, ``misc.memprof.start()`` returns ``true``.
 
-Stopping profiler in Lua code:
+The Lua code for stopping the profiler -- as in line 18 in the test.lua example above -- is:
```
```suggestion
The Lua code for stopping the profiler --- as in line 18 in the test.lua example above --- is:
```

NickVolynkin commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -61,38 +61,40 @@ Below is a chunk of simple Lua code named ``test.lua`` to illustrate this.
-a system-dependent error code as the third result.
-Otherwise, it returns ``true``.
+If the operation fails,
+for example if  there is an error when the file descriptor is being closed
```
```suggestion
for example if there is an error when the file descriptor is being closed
```

NickVolynkin commented:
@pgulutzan I've just fixed a couple of syntax errors and rebased this on the `latest` branch. Again, I apologize for not doing it earlier.

ladyune commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -61,38 +61,40 @@ Below is a chunk of simple Lua code named ``test.lua`` to illustrate this.
+``misc.memprof.stop()`` returns ``nil`` as the first result,
+an error-message string as the second result,
+and a system-dependent error code number as the third result.
+If the operation succeeds, ``misc.memprof.start()`` returns ``true``.
```
Should it be ``misc.memprof.stop()`` here?

patiencedaur replied:
@pgulutzan Kindly help us resolve this issue. Is there a typo here? Thank you in advance.

patiencedaur replied:
```suggestion
If the operation succeeds, ``misc.memprof.stop()`` returns ``true``.
```

patiencedaur commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -14,15 +14,15 @@ that put the most pressure on the Lua garbage collector (GC).
+Usage of the profiler involves two steps:
 
-1.  :ref:`Collect <profiler_usage_get>` a binary profile of allocations,
+1.  :ref:`Collecting <profiler_usage_get>` a binary profile of allocations,
```
@Buristan binary profile = бинарный или двоичный профилировщик?

Buristan replied:
Бинарный профиль -- то, что напрофилировали.

patiencedaur commented:
@Buristan 
[Lines 33-36](https://github.com/tarantool/doc/pull/2251/files#diff-9af01d5422ec57beeeb70a8daa3a3231f4f457e4b7145439ebed8d49dc210138R33-R36): 

1. какую часть примера нужно поместить между двумя функциями misc.memprof — т.е. что имеется в виду под this part?
2. Prevent allocations on traces - предотвращение аллокации при трассировке?
3. frame - стековый фрейм?

patiencedaur commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -61,38 +61,40 @@ Below is a chunk of simple Lua code named ``test.lua`` to illustrate this.
     local str, err = misc.memprof.start(FILENAME)
 
-where ``FILENAME`` is a name of the binary file where profiling events are written.
+where ``FILENAME`` is the name of the binary file where profiling events are written.
```
@Buristan
profiling events = the events that are profiled? i.e. not the events that happen during profiling?

Buristan replied:
Profiling events are described above. There are 3 possible events interesting for memory profiler: allocation, reallocation and deallocation.

Buristan commented:
> @Buristan [Строки 33-63](https://github.com/tarantool/doc/pull/2251/files#diff-9af01d5422ec57beeeb70a8daa3a3231f4f457e4b7145439ebed8d49dc210138L33-L63); [фрагмент на тестовом сайте](https://develop.tarantool.io/en/doc/pgulutzan-gh-2113/book/app_server/luajit_memprof/#collecting-binary-profile).
> 
>     1. здесь приведён пример кода. какую часть этого примера нужно поместить между двумя функциями misc.memprof — т.е. что имеется в виду под this part?
> 

Тот кусок кода, который пользователь хочет запрофилировать. То есть для примера из доки это следующий код:
```lua
local function append(str, rep)
    return string.rep(str, rep)
end

local t = {}
for i = 1, 1e4 do
    -- table.insert is the built-in function and all corresponding
    -- allocations are reported in the scope of the main chunk.
    table.insert(t,
        append('q', i)
    )
end```

>     2. Prevent allocations on traces - предотвращение аллокации при трассировке?

Скорее "избегание" аллокаций на трассах (куски скомпилированного луа кода). Мемпроф пока что все аллокации, которые произошли на трассах, записывает как INTERNAL -- поэтому лучше запускать профилировщик с выключенным JIT - ом, чтобы можно было подробнее посмотреть, что происходит.

> 
>     3. frame - стековый фрейм?

Да, фрейм Lua стека.

patiencedaur commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -514,15 +553,42 @@ the profiling report becomes much prettier:
+
+The end of each display is a HEAP SUMMARY section which looks like this:
+
+@<filename>:<line number> holds <number of still reachable bytes> bytes: <number of allocation events> allocs, <number of deallocation events> frees
```
@NickVolynkin 
Should this be a code block?

patiencedaur reviewed (commented):
Fix a typo

patiencedaur commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -201,23 +201,22 @@ An event record has the following format:
 
 The ``Overrides`` label shows what allocation has been overridden.
 
+See the :ref:`tarantool test.lua chunk above<profiler_usage_example01>`
```
```suggestion
See the :ref:`test.lua chunk above<profiler_usage_example01>`
```
The link points to the contents of `test.lua`, not to the `tarantool test.lua` command.

patiencedaur reviewed (commented):
Minor inconsistency

patiencedaur commented:
@pgulutzan I've seen you approve similar style-related suggestions in other pull requests. Would you mind if I committed stylistic changes right away, without waiting for your approval?

pgulutzan commented on doc/book/app_server/luajit_memprof.rst:
```diff
@@ -201,23 +201,22 @@ An event record has the following format:
 ..  code-block:: text
 
-    @<filename>:<function_line>, line <line_number>: <number_of_events> events +<allocated> bytes -<freed> bytes
+    @<filename><line_number>: <number_of_events> events +<allocated> bytes -<freed> bytes
```
That is correct, I should have used a colon there, thanks.
